File: src/main/java/features/reports/pdf/PdfElementsAccumulator.java
package features.reports.pdf;

import com.itextpdf.kernel.geom.PageSize;
import com.itextpdf.kernel.pdf.PdfDocument;
import com.itextpdf.kernel.pdf.PdfWriter;
import com.itextpdf.layout.Document;
import core.algorithm.Algorithm;
import core.data.CollectionData;
import core.data.FileData;
import features.reports.PdfElementData;

import java.io.ByteArrayOutputStream;

/**
 * {@link Algorithm} that puts all {@link PdfElementData} from input to PDF document and returns {@link FileData} with generated PDF
 */
public class PdfElementsAccumulator implements Algorithm<CollectionData<PdfElementData>, FileData> {

    private static final String EXTENSION = ".pdf";
    private String reportName;

    public PdfElementsAccumulator(String reportName) {
        this.reportName = reportName;
    }

    @Override
    public FileData execute(CollectionData<PdfElementData> elements) {
        var resultStream = new ByteArrayOutputStream();
        var writer = new PdfWriter(resultStream);

        var pdf = new PdfDocument(writer);

        var document = new Document(pdf, PageSize.A4);

        document.setMargins(10, 10, 10, 10);

        for (var element : elements.getCollection())
            document.add(element.pdfElement);

        document.close();
        return new FileData(resultStream.toByteArray(), reportName, EXTENSION);
    }
}

File: src/main/kotlin/com/mirf/features/pdf/PdfElementsAccumulator.kt
package com.mirf.features.pdf

import com.itextpdf.kernel.geom.PageSize
import com.itextpdf.kernel.pdf.PdfDocument
import com.itextpdf.kernel.pdf.PdfWriter
import com.itextpdf.layout.Document
import com.mirf.core.algorithm.Algorithm
import com.mirf.core.data.CollectionData
import com.mirf.core.data.FileData
import com.mirf.features.reports.PdfElementData

import java.io.ByteArrayOutputStream

/**
 * [Algorithm] that puts all [PdfElementData] from input to PDF document and returns [FileData] with generated PDF
 */
class PdfElementsAccumulator(private val reportName: String) : Algorithm<CollectionData<PdfElementData>, FileData> {

    override fun execute(input: CollectionData<PdfElementData>): FileData {

        val resultStream = ByteArrayOutputStream()
        val writer = PdfWriter(resultStream)

        val pdf = PdfDocument(writer)

        val document = Document(pdf, PageSize.A4)

        document.setMargins(10f, 10f, 10f, 10f)

        for (element in input.collection)
            document.add(element.pdfElement)

        document.close()
        return FileData(resultStream.toByteArray(), reportName, EXTENSION)
    }

    companion object {

        private const val EXTENSION = ".pdf"
    }
}
